cmd/src/server.rs

```diff
@@ -1,442 +1,587 @@
-
-    let bps = i64::try_from(cfg.server.snap_max_write_bytes_per_sec.0)
-        .unwrap_or_else(|_| fatal!("snap_max_write_bytes_per_sec > i64::max_value"));
-
-    // Create snapshot manager, server.
-    let snap_mgr = SnapManagerBuilder::default()
-        .max_write_bytes_per_sec(bps)
-        .max_total_size(cfg.server.snap_max_total_size.0)
-        .build(
-            snap_path.as_path().to_str().unwrap().to_owned(),
-            Some(router.clone()),
+        let security_mgr =
+            Arc::new(SecurityManager::new(&config.security).unwrap_or_else(|e| {
+                fatal!("failed to create security manager: {}", e.description())
+            }));
+
+        let pd_client = Arc::new(
+            RpcClient::new(&config.pd, security_mgr.clone())
+                .unwrap_or_else(|e| fatal!("failed to create rpc client: {}", e)),
         );
 
-    let server_cfg = Arc::new(cfg.server.clone());
-
-    // Create coprocessor endpoint.
-    let cop_read_pool = coprocessor::readpool_impl::build_read_pool(
-        &cfg.readpool.coprocessor,
-        pd_sender.clone(),
-        engine.clone(),
-    );
-    let cop = coprocessor::Endpoint::new(&server_cfg, cop_read_pool);
-
-    let importer = Arc::new(SSTImporter::new(import_path).unwrap());
-    let import_service = ImportSSTService::new(
-        cfg.import.clone(),
-        raft_router.clone(),
-        engines.kv.clone(),
-        Arc::clone(&importer),
-    );
-
-    // The `DebugService` and `DiagnosticsService` will share the same thread pool
-    let pool = Builder::new()
-        .name_prefix(thd_name!("debugger"))
-        .pool_size(1)
-        .create();
-    // Create Debug service.
-    let debug_service = DebugService::new(
-        engines.clone(),
-        pool.clone(),
-        raft_router.clone(),
-        gc_worker.clone(),
-    );
-
-    // Create Diagnostics service
-    let diag_service = DiagnosticsService::new(pool, cfg.log_file.clone());
-
-    // Create Backup service.
-    let mut backup_worker = tikv_util::worker::Worker::new("backup-endpoint");
-    let backup_scheduler = backup_worker.scheduler();
-    let backup_service = backup::Service::new(backup_scheduler);
-
-    // Create server
-    let mut server = Server::new(
-        &server_cfg,
-        &security_mgr,
-        storage.clone(),
-        cop,
-        raft_router,
-        resolver.clone(),
-        snap_mgr.clone(),
-        gc_worker.clone(),
-    )
-    .unwrap_or_else(|e| fatal!("failed to create server: {}", e));
-
-    // Register services.
-    if server
-        .register_service(create_import_sst(import_service))
-        .is_some()
-    {
-        fatal!("failed to register import service");
-    }
-    if server
-        .register_service(create_debug(debug_service))
-        .is_some()
-    {
-        fatal!("failed to register debug service");
+        let (resolve_worker, resolver) = resolve::new_resolver(Arc::clone(&pd_client))
```

Comment:
Cluster ID should be checked first before doing anything.

Reply:
Why? The only work that is done is local initialisation, afaict we don't communicate with the network. If the cluster id check fails, then we quit, so it seems like it should be fine to check the cluster id at the start of `connect_to_pd`

Reply:
Because all other operations are implemented under the assumption that the store belongs to the cluster. Delaying the check certainly expose space for bugs and gotchas.

Reply:
OK, I've addressed this with the latest commit. @BusyJay PTAL :-)